pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,353 @@
+		routeClient: routeClient,
+		namespace:   clusterNS,
+		nodeMap:     make(map[string]*v1.Node),
+		ownerRefs:   ownerRefs,
+	}, nil
+}
+
+// ListRoutes implements Routes.ListRoutes
+// Get RouteSet CR from SC namespace and then filters routes that belong to the specified clusterName
+// Only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+func (r *routesProvider) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
+	klog.V(6).Infof("Attempting to list Routes for cluster %s", clusterName)
+
+	// use labelSelector to filter RouteSet CRs that belong to this cluster
+	labelSelector := metav1.LabelSelector{
+		MatchLabels: map[string]string{LabelKeyClusterName: clusterName},
+	}
+	routeSets, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String(),
+	})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return []*cloudprovider.Route{}, nil
+		}
+		klog.ErrorS(ErrListRouteSet, fmt.Sprintf("%v", err))
+		return nil, err
+	}
+	if len(routeSets.Items) == 0 {
+		return []*cloudprovider.Route{}, nil
+	}
+	return r.createCPRoutes(routeSets), nil
+}
+
+// createCPRoutes creates cloudprovider Routes based on RouteSet CR
+func (r *routesProvider) createCPRoutes(routeSets *v1alpha1.RouteSetList) []*cloudprovider.Route {
+	var routes []*cloudprovider.Route
+	for _, routeSet := range routeSets.Items {
+		// only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+		condition := GetRouteSetCondition(&(routeSet.Status), v1alpha1.RouteSetConditionTypeReady)
+		if condition != nil && condition.Status == v1.ConditionTrue {
+			// one RouteSet per node, so we can use nodeName as the name of RouteSet CR
+			nodeName := routeSet.Name
+			for _, route := range routeSet.Spec.Routes {
+				cpRoute := &cloudprovider.Route{
+					Name:            route.Name,
+					TargetNode:      types.NodeName(nodeName),
+					DestinationCIDR: route.Destination,
+				}
+				routes = append(routes, cpRoute)
+			}
+		}
+	}
+	return routes
+}
+
+// CreateRoute implements Routes.CreateRoute
+// Create a RouteSet custom resource for a Node
+func (r *routesProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	nodeName := string(route.TargetNode)
+	klog.V(6).Infof("Creating Route for node %s with hint %s in cluster %s", nodeName, nameHint, clusterName)
+
+	nodeIP, err := r.getNodeIPAddress(nodeName, util.IsIPv4(route.DestinationCIDR))
+	if err != nil {
+		klog.Errorf("getting node %s IP address failed: %v", nodeName, err)
+		return err
+	}
+
+	routeSet, err := r.createRouteSetCR(ctx, clusterName, nameHint, nodeName, route.DestinationCIDR, nodeIP)
+	if err != nil {
+		klog.Errorf("creating RouteSet CR for node %s failed: %s", nodeName, err)
+		return err
+	}
+	// check realized state of static routes
+	return r.checkStaticRouteRealizedState(routeSet.Name)
```

Comment:
Do we need to check realized state after creating RouteSet CR here? I mean cloud provider returns realized routes in `ListRoutes` interface

Reply:
if we don't check realized state here, it's possible that we've created a routeSet CR, but the static route is not actually being created
in that case, next time when we do `ListRoutes`, one route is missing for a node, then we will need to try `CreateRoute` one more time: https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go#L194
but if we check the realized state at this point, we could retry as well: https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go#L188
But I agree that either way works, cuz the route controller does retry within a few seconds